saw-core/src/Verifier/SAW/Term/Pretty.hs

```diff
@@ -289,22 +294,43 @@ withBoundVarM basename m =
                               ppLocalMemoTable = IntMap.empty }) m
      return (var, ret)
 
--- | Run a computation in the context of a fresh "memoization variable" that is
--- bound to the given term index, passing the new memoization variable to the
--- computation. If the flag is true, use the global table, otherwise use the
--- local table.
-withMemoVar :: Bool -> TermIndex -> (MemoVar -> PPM a) -> PPM a
+-- | Attempt to memoize the given term (index) 'idx' and run a computation in
+-- the context that the attempt produces. If memoization succeeds, the context
+-- will contain a binding (global in scope if 'global_p' is set, local if not)
+-- of a fresh memoization variable to the term, and the fresh variable will be
+-- supplied to the computation. If memoization fails, the context will not
+-- contain such a binding, and no fresh variable will be supplied.
+withMemoVar :: Bool -> TermIndex -> (Maybe MemoVar -> PPM a) -> PPM a
 withMemoVar global_p idx f =
-  do memo_var <- ppNextMemoVar <$> ask
-     local (\s -> add_to_table global_p memo_var s) (f memo_var)
-       where
-         add_to_table True v st =
-           st { ppNextMemoVar = v + 1,
-                ppGlobalMemoTable = IntMap.insert idx v (ppGlobalMemoTable st) }
-         add_to_table False v st =
-           st { ppNextMemoVar = v + 1,
-                ppLocalMemoTable = IntMap.insert idx v (ppLocalMemoTable st) }
+  do
+    memoVar <- asks ppNextMemoVar
+    memoSkips <- asks (ppNoInlineMemo . ppOpts)
+    idxSkips <- asks (ppNoInlineIdx . ppOpts)
+    case memoSkips of
+      -- Even if we must skip this memoization variable, we still want to
+      -- "pretend" we memoized by calling `updateMemoVar`, so that non-inlined
+      -- memoization identifiers are kept constant between two
+      -- otherwise-identical terms with differing inline strategies.
+      (skip:_) | skip == memoVar -> local (updateMemoVar . addIdxSkip . removeMemoSkip) (f Nothing)
+      _ | idx `Set.member` idxSkips -> f Nothing
+      _ -> local (updateMemoVar . bind memoVar) (f (Just memoVar))
+  where
+    bind = if global_p then bindGlobal else bindLocal
+
+    bindGlobal memoVar PPState{ .. } =
+      PPState { ppGlobalMemoTable = IntMap.insert idx memoVar ppGlobalMemoTable, .. }
+
+    bindLocal memoVar PPState{ .. } =
+      PPState { ppLocalMemoTable = IntMap.insert idx memoVar ppLocalMemoTable, .. }
+
+    removeMemoSkip PPState{ ppOpts = PPOpts{ .. }, .. } =
+      PPState { ppOpts = PPOpts { ppNoInlineMemo = tail ppNoInlineMemo, .. }, .. }
```

Comment:
`tail` is a partial function, which makes me nervous. In this particular circumstance, this use is safe due to the fact that `removeMemoSkip` is called in a case where `ppNoInlineMemo` is known to be non-empty.

That being said, you could avoid the use of `tail` entirely by passing the memo skips to this function as an argument. That is, do something like this:

```hs
      (skip:skips) | skip == memoVar -> local (updateMemoVar . addIdxSkip . setMemoSkips skips) (f Nothing)

...

    setMemoSkips memoSkips PPState{ ppOpts = PPOpts{ .. }, .. } =
      PPState { ppOpts = PPOpts { ppNoInlineMemo = memoSkips, .. }, .. }
```

Reply:
Good idea.